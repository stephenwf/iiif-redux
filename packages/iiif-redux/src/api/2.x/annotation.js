import memoize from 'lodash.memoize';
import { createSelector, createStructuredSelector } from 'reselect';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as structural from './iiif/structural';
import {
  getAllCanvases,
  getAllExternalResources,
  getAllImages,
  getAllLayers,
  getAllServices,
} from '../all';
import * as linking from './iiif/linking';
import { isImageService } from '../../constants/services';
import { standardFieldMappingFactory } from '../../utility/new/standardFieldMappingFactory';
const annotation = memoize(selector => {
  /**************************************************
   * Technical properties
   *
   * - getId
   * - getType
   * - getViewingHint
   * - getMotivation
   **************************************************/
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingHint = createSelector(
    selector,
    technical.getWhitelistedViewingHint([])
  );

  const getMotivation = createSelector(selector, technical.getMotivation);

  /**************************************************
   * Descriptive properties
   *
   * - getLabel
   * - getMetadata
   * - getDescription
   * - getThumbnail
   * - getAttribution
   * - getLicense
   * - getLogo
   **************************************************/
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getDescription = createSelector(selector, descriptive.getDescription);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getAttribution = createSelector(selector, descriptive.getAttribution);

  const getLogoIds = createSelector(selector, descriptive.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllImages,
    (logoIds, allImages) => {
      return logoIds.map(logoId => allImages[logoId] || logoId);
    }
  );

  const getLicense = createSelector(selector, descriptive.getLicense);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);
  const getThumbnail = createSelector(
    getThumbnailId,
    getAllImages,
    (thumbnailId, allImages) => allImages[thumbnailId] || thumbnailId
  );

  /**************************************************
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getRelated
   * - getRendering
   * - getWithin
   **************************************************/
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllExternalResources,
    (seeAlsoIds, allExternalResources) =>
      seeAlsoIds.map(
        seeAlsoId =>
          allExternalResources[seeAlsoId] || {
            '@id': seeAlsoId,
            label: [{ '@value': 'unknown', '@language': '@none' }],
          }
      )
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(
    getServiceIds,
    getAllServices,
    (serviceIds, allServices) =>
      serviceIds.map(serviceId => allServices[serviceId])
  );

  const getRelatedIds = createSelector(selector, linking.getRelated);
  const getRelated = createSelector(
    getRelatedIds,
    getAllExternalResources,
    (relatedIds, allExternalResources) =>
      relatedIds.map(relatedId => allExternalResources[relatedId])
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllExternalResources,
    (renderingIds, allExternalResources) =>
      renderingIds.map(renderingId => allExternalResources[renderingId])
  );

  const getWithinIds = createSelector(selector, linking.getWithin);
  const getWithin = createSelector(
    getWithinIds,
    getAllLayers,
    getAllExternalResources,
    (withinIds, allLayers, allExternalResources) =>
      withinIds
        .map(within => allLayers[within.id] || allExternalResources[within.id])
        .filter(e => e)
  );

  /**************************************************
   * Structural properties
   *
   * - getResource
   **************************************************/
  const getResourceId = createSelector(selector, structural.getResource);
  const getResource = createSelector(
    getResourceId,
    getAllImages,
    (resourceId, allImageResources) => allImageResources[resourceId] || null
  );

  const getImageService = createSelector(
    getResource,
    getAllServices,
    (resource, allServices) =>
      resource
        ? resource.service.reduce((innerResult, serviceId) => {
            if (innerResult) return innerResult;
            const service = allServices[serviceId];
            if (isImageService(service.profile)) {
              return service;
            }
            return null;
          }, null)
        : null
  );

  const getOnId = createSelector(selector, resource => resource.on);
  const getOn = createSelector(
    getOnId,
    getAllCanvases,
    (canvasId, canvases) => canvases[canvasId] || null
  );

  return {
    // Technical
    getId,
    getType,
    getViewingHint,
    getMotivation,
    // Descriptive
    getLabel,
    getMetadata,
    getDescription,
    getThumbnailId,
    getThumbnail,
    getAttribution,
    getLicense,
    getLogoIds,
    getLogo,
    // Linking
    getSeeAlso,
    getSeeAlsoIds,
    getService,
    getServiceIds,
    getRelated,
    getRelatedIds,
    getRendering,
    getRenderingIds,
    getWithin,
    getWithinIds,
    // Structural
    getResourceId,
    getResource,
    // Service
    getImageService,
    // Annotation specifics
    getOn,
    getOnId,
  };
});

export const mappings = standardFieldMappingFactory(annotation);

export default annotation;
