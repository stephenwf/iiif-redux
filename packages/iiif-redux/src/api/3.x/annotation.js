import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationPages,
  getAllCanvases,
  getAllCanvasReferences,
  getAllContentResources,
  getAllExternalResources,
  getAllRanges,
  getAllResources,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';

const annotation = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getBehavior
   * - getTimeMode
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getBehavior = createSelector(selector, technical.getBehavior);

  const getTimeMode = createSelector(selector, technical.getTimeMode);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getRequiredStatement
   * - getRights
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllExternalResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllExternalResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllExternalResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  return {
    getId,
    getType,
    getTimeMode,
    getBehavior,
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getRequiredStatement,
    getRights,
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
  };
});

export default annotation;
