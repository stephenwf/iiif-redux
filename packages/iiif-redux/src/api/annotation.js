import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import presentation2Annotation from './2.x/annotation';
import { getAllAnnotations } from './all';

const annotation = presentation2Annotation;

export default annotation;

export const annotations = resourceListSelectorFactory(
  getAllAnnotations,
  annotation
);

export const annotationByIdSelector = byIdSelectorFactory(
  annotation,
  'annotations'
);
