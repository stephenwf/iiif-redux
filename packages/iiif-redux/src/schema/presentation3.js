import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';

function createEntity(name) {
  const options = {
    idAttribute: 'id',
  };

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations'); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollection'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReference'); // 10

// Unofficial types.
const externalResource = createEntity('externalResources');
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    externalResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'externalResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    canvasReference,
    canvas,
  },
  entity => {
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBody = new schema.Union(
  {
    contentResource,
    externalResource,
    choice,
  },
  input => {
    switch (input.type) {
      case 'Application':
      case 'Dataset':
      case 'Image':
      case 'Sound':
      case 'Text':
      case 'Video':
      case 'TextualBody':
        return 'contentResource';
      case 'Choice':
        return 'choice';

      // @todo more content to add.
      default:
        return 'externalResource';
    }
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    range,
    canvasReference,
  },
  input => {
    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: annotationBody,
  target: canvas,

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  rendering: [externalResource],
  partOf: partOf,
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  externalResource,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
