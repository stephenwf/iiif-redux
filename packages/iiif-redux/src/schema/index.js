import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
