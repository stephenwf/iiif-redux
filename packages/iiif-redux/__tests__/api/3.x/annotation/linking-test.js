import annotation from '../../../../src/api/3.x/annotation';
import { normalize } from '../../../../src/schema/presentation3';
import fixtures from './fixtures';
import { DEFAULT_STATE } from '../../../../src/spaces/iiif-resource';

describe('api/3.x/annotation/linking', () => {
  fixtures.forEach(({ fixture, name }) => {
    describe(`Annotation: ${name}`, () => {
      const {
        result: { id },
        entities,
      } = normalize(fixture);
      const state = {
        resources: { ...DEFAULT_STATE.resources, ...entities },
      };
      const selectAnnotation = s => s.resources.annotations[id];

      const {
        getSeeAlsoIds,
        getSeeAlso,
        getServiceIds,
        getService,
        getLogoIds,
        getLogo,
        getHomepageId,
        getHomepage,
        getRenderingIds,
        getRendering,
        getPartOfId,
        getPartOf,
        getBodyId,
        getBody,
        getTargetId,
        getTarget,
      } = annotation(selectAnnotation);

      test('getSeeAlsoIds', () => {
        expect(getSeeAlsoIds(state)).toMatchSnapshot(`getSeeAlsoIds ${name}`);
      });

      test('getSeeAlso', () => {
        expect(getSeeAlso(state)).toMatchSnapshot(`getSeeAlso ${name}`);
      });

      test('getServiceIds', () => {
        expect(getServiceIds(state)).toMatchSnapshot(`getServiceIds ${name}`);
      });

      test('getService', () => {
        expect(getService(state)).toMatchSnapshot(`getService ${name}`);
      });

      test('getLogoIds', () => {
        expect(getLogoIds(state)).toMatchSnapshot(`getLogoIds ${name}`);
      });

      test('getLogo', () => {
        expect(getLogo(state)).toMatchSnapshot(`getLogo ${name}`);
      });

      test('getHomepageId', () => {
        expect(getHomepageId(state)).toMatchSnapshot(`getHomepageId ${name}`);
      });

      test('getHomepage', () => {
        expect(getHomepage(state)).toMatchSnapshot(`getHomepage ${name}`);
      });

      test('getRenderingIds', () => {
        expect(getRenderingIds(state)).toMatchSnapshot(
          `getRenderingIds ${name}`
        );
      });

      test('getRendering', () => {
        expect(getRendering(state)).toMatchSnapshot(`getRendering ${name}`);
      });

      test('getPartOfId', () => {
        expect(getPartOfId(state)).toMatchSnapshot(`getPartOfId ${name}`);
      });

      test('getPartOf', () => {
        expect(getPartOf(state)).toMatchSnapshot(`getPartOf ${name}`);
      });

      test('getBodyId', () => {
        expect(getBodyId(state)).toMatchSnapshot(`getBodyId ${name}`);
      });

      test('getBody', () => {
        expect(getBody(state)).toMatchSnapshot(`getBody ${name}`);
      });

      test('getTargetId', () => {
        expect(getTargetId(state)).toMatchSnapshot(`getTargetId ${name}`);
      });

      test('getTarget', () => {
        expect(getTarget(state)).toMatchSnapshot(`getTarget ${name}`);
      });
    });
  });
});
