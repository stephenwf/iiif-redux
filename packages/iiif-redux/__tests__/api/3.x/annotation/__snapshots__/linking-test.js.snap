// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/3.x/annotation/linking Annotation: Choice body example getBody: getBody Choice body example 1`] = `
Array [
  Object {
    "default": "http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png",
    "id": "https://example.org/iiif/choice/1",
    "item": Array [
      "http://iiif.io/api/image/2.1/example/reference/page1-full/full/full/0/gray.jpg",
    ],
    "type": "Choice",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getBodyId: getBodyId Choice body example 1`] = `
Array [
  Object {
    "id": "https://example.org/iiif/choice/1",
    "schema": "choice",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getHomepage: getHomepage Choice body example 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getHomepageId: getHomepageId Choice body example 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getLogo: getLogo Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getLogoIds: getLogoIds Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getPartOf: getPartOf Choice body example 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getPartOfId: getPartOfId Choice body example 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getRendering: getRendering Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getRenderingIds: getRenderingIds Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getSeeAlso: getSeeAlso Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getSeeAlsoIds: getSeeAlsoIds Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getService: getService Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getServiceIds: getServiceIds Choice body example 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getTarget: getTarget Choice body example 1`] = `
Array [
  "https://example.org/iiif/book1/canvas/p1",
]
`;

exports[`api/3.x/annotation/linking Annotation: Choice body example getTargetId: getTargetId Choice body example 1`] = `
Array [
  "https://example.org/iiif/book1/canvas/p1",
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getBody: getBody Complete annotation 1`] = `
Array [
  Object {
    "format": "image/jpeg",
    "id": "https://example.org/iiif/book1/res/page1.jpg",
    "type": "Image",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getBodyId: getBodyId Complete annotation 1`] = `
Array [
  Object {
    "id": "https://example.org/iiif/book1/res/page1.jpg",
    "schema": "contentResource",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getHomepage: getHomepage Complete annotation 1`] = `
Object {
  "format": "text/html",
  "id": "https://example.org/info/book1/",
  "label": Object {
    "en": Array [
      "Home page for Book 1",
    ],
  },
  "type": "Text",
}
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getHomepageId: getHomepageId Complete annotation 1`] = `"https://example.org/info/book1/"`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getLogo: getLogo Complete annotation 1`] = `
Array [
  Object {
    "id": "https://example.org/logos/institution1.jpg",
    "service": Array [
      "https://example.org/service/inst1",
    ],
    "type": "Image",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getLogoIds: getLogoIds Complete annotation 1`] = `
Array [
  "https://example.org/logos/institution1.jpg",
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getPartOf: getPartOf Complete annotation 1`] = `
Array [
  Object {
    "id": "https://example.org/collections/books/",
    "type": "Collection",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getPartOfId: getPartOfId Complete annotation 1`] = `
Array [
  Object {
    "id": "https://example.org/collections/books/",
    "schema": "collection",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getRendering: getRendering Complete annotation 1`] = `
Array [
  Object {
    "format": "application/pdf",
    "id": "https://example.org/iiif/book1.pdf",
    "label": Object {
      "en": Array [
        "Download as PDF",
      ],
    },
    "type": "Text",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getRenderingIds: getRenderingIds Complete annotation 1`] = `
Array [
  "https://example.org/iiif/book1.pdf",
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getSeeAlso: getSeeAlso Complete annotation 1`] = `
Array [
  Object {
    "format": "text/xml",
    "id": "https://example.org/library/catalog/book1.xml",
    "profile": "https://example.org/profiles/bibliographic",
    "type": "Dataset",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getSeeAlsoIds: getSeeAlsoIds Complete annotation 1`] = `
Array [
  "https://example.org/library/catalog/book1.xml",
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getService: getService Complete annotation 1`] = `
Array [
  Object {
    "id": "https://example.org/service/example",
    "profile": "https://example.org/docs/example-service.html",
    "type": "Service",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getServiceIds: getServiceIds Complete annotation 1`] = `
Array [
  "https://example.org/service/example",
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getTarget: getTarget Complete annotation 1`] = `
Array [
  "https://example.org/iiif/book1/canvas/p1",
]
`;

exports[`api/3.x/annotation/linking Annotation: Complete annotation getTargetId: getTargetId Complete annotation 1`] = `
Array [
  "https://example.org/iiif/book1/canvas/p1",
]
`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getBody: getBody IIIF Documentation fixture 1`] = `
Array [
  Object {
    "format": "image/jpeg",
    "id": "https://example.org/iiif/book1/res/page1.jpg",
    "type": "Image",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getBodyId: getBodyId IIIF Documentation fixture 1`] = `
Array [
  Object {
    "id": "https://example.org/iiif/book1/res/page1.jpg",
    "schema": "contentResource",
  },
]
`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getHomepage: getHomepage IIIF Documentation fixture 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getHomepageId: getHomepageId IIIF Documentation fixture 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getLogo: getLogo IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getLogoIds: getLogoIds IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getPartOf: getPartOf IIIF Documentation fixture 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getPartOfId: getPartOfId IIIF Documentation fixture 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getRendering: getRendering IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getRenderingIds: getRenderingIds IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getSeeAlso: getSeeAlso IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getSeeAlsoIds: getSeeAlsoIds IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getService: getService IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getServiceIds: getServiceIds IIIF Documentation fixture 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getTarget: getTarget IIIF Documentation fixture 1`] = `
Array [
  "https://example.org/iiif/book1/canvas/p1",
]
`;

exports[`api/3.x/annotation/linking Annotation: IIIF Documentation fixture getTargetId: getTargetId IIIF Documentation fixture 1`] = `
Array [
  "https://example.org/iiif/book1/canvas/p1",
]
`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getBody: getBody Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getBodyId: getBodyId Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getHomepage: getHomepage Minimal annotation 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getHomepageId: getHomepageId Minimal annotation 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getLogo: getLogo Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getLogoIds: getLogoIds Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getPartOf: getPartOf Minimal annotation 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getPartOfId: getPartOfId Minimal annotation 1`] = `null`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getRendering: getRendering Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getRenderingIds: getRenderingIds Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getSeeAlso: getSeeAlso Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getSeeAlsoIds: getSeeAlsoIds Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getService: getService Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getServiceIds: getServiceIds Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getTarget: getTarget Minimal annotation 1`] = `Array []`;

exports[`api/3.x/annotation/linking Annotation: Minimal annotation getTargetId: getTargetId Minimal annotation 1`] = `Array []`;
